access/src/main/java/org/apache/accumulo/access/AccessExpressionImpl.java

```diff
@@ -0,0 +1,635 @@
+
+  // must create this after creating EMPTY_NODE
+  static final AccessExpression EMPTY = new AccessExpressionImpl("");
+
+  /**
+   * A node in the parse tree for a visibility expression.
+   */
+  static class Node {
+    /**
+     * An empty list of nodes.
+     */
+    public static final List<Node> EMPTY = Collections.emptyList();
+    NodeType type;
+    int start;
+    int end;
+    List<Node> children = EMPTY;
+
+    public Node(NodeType type, int start) {
+      this.type = type;
+      this.start = start;
+      this.end = start + 1;
+    }
+
+    public Node(int start, int end) {
+      this.type = NodeType.TERM;
+      this.start = start;
+      this.end = end;
+    }
+
+    public void add(Node child) {
+      if (children == EMPTY) {
+        children = new ArrayList<>();
+      }
+
+      children.add(child);
+    }
+
+    public NodeType getType() {
+      return type;
+    }
+
+    public List<Node> getChildren() {
+      return children;
+    }
+
+    public BytesWrapper getTerm(byte[] expression) {
+      if (type != NodeType.TERM) {
+        throw new IllegalStateException();
+      }
+
+      if (expression[start] == '"') {
+        // its a quoted term
+        int qStart = start + 1;
+        int qEnd = end - 1;
+
+        return new BytesWrapper(expression, qStart, qEnd - qStart);
+      }
+      return new BytesWrapper(expression, start, end - start);
+    }
+  }
+
+  /**
+   * A node comparator. Nodes sort according to node type, terms sort lexicographically. AND and OR
+   * nodes sort by number of children, or if the same by corresponding children.
+   */
+  static class NodeComparator implements Comparator<Node>, Serializable {
+
+    private static final long serialVersionUID = 1L;
+    byte[] text;
+
+    /**
+     * Creates a new comparator.
+     *
+     * @param text expression string, encoded in UTF-8
+     */
+    public NodeComparator(byte[] text) {
+      this.text = text;
+    }
+
+    @Override
+    public int compare(Node a, Node b) {
+      int diff = a.type.ordinal() - b.type.ordinal();
+      if (diff != 0) {
+        return diff;
+      }
+      switch (a.type) {
+        case EMPTY:
+          return 0; // All empty nodes are the same
+        case TERM:
+          return Arrays.compare(text, a.start, a.end, text, b.start, b.end);
+        case OR:
+        case AND:
+          diff = a.children.size() - b.children.size();
+          if (diff != 0) {
+            return diff;
+          }
+          for (int i = 0; i < a.children.size(); i++) {
+            diff = compare(a.children.get(i), b.children.get(i));
+            if (diff != 0) {
+              return diff;
+            }
+          }
+      }
+      return 0;
+    }
+  }
+
+  /*
+   * Convenience method that delegates to normalize with a new NodeComparator constructed using the
+   * supplied expression.
+   */
+  public static Node normalize(Node root, byte[] expression) {
```

Comment:
Is this still needed?

Reply:
yeah its used by the impl of AccessExpression.normalize()